core/block_validator.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/pow"
+	"gopkg.in/fatih/set.v0"
+)
+
+type BlockValidator struct {
+	// Canonical block chain
+	bc *BlockChain
+	// Proof of work used for validating
+	Pow pow.PoW
+}
+
+// NewBlockValidator returns a new block validator which is safe for re-use
+func NewBlockValidator(blockchain *BlockChain, pow pow.PoW) *BlockValidator {
+	sm := &BlockValidator{
+		Pow: pow,
+		bc:  blockchain,
+	}
+	return sm
+}
+
+func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+	if v.bc.HasBlock(block.Hash()) {
+		return &KnownBlockError{block.Number(), block.Hash()}
+	}
```

Comment:
I think this check will not be enough as of the fast sync. A block may exist but the associated state trie not (i.e. if the block is part of the fast-sync phase). Not checking the state root may/will cause block processing issues where a block content (header + body) is present, but the state is not: the processor will skip the import as it thinks the block's already known, but processing the next block will fail as the state trie wasn't known after all.

A sample scenario where this actually occurred:
 * Start fast sync, pull in various headers, block bodies, receipts.
 * At the very end, sync fails. It's assumed to be an attack (at least an unsafe scenario), fast sync disables itself.
 * Slow sync starts from scratch since no block was properly imported.
 * Block processor skips all the initial ones, since they are already present.
 * When a fully new block is found, import fails since the previous state trie was never assembled.

(This scenario actually happened, hence why these additions were made: https://github.com/ethereum/go-ethereum/pull/1917/files#diff-9dc4dfc538cf66079b5ddd793724e148L215 and https://github.com/ethereum/go-ethereum/pull/1917/files#diff-9dc4dfc538cf66079b5ddd793724e148L220, both important).